numpy/core/src/multiarray/ctors.c

```diff
@@ -816,6 +840,23 @@ PyArray_NewFromDescr_int(
             goto fail;
         }
 
+        /*
+         * If the array needs special dtype-specific zero-filling logic, do that
+         */
+        if (zeroed && (fill_zero_info.func != NULL)) {
```

Comment:
```suggestion
        if (NPY_UNLIKELY(zeroed && (fill_zero_info.func != NULL))) {
```
Probably doesn't matter, but we could flag it as unlikely (the effect is that the compiler will move this code to the end, so if it isn't hit it can continue quicker with what follows).